core/src/main/scala/org/apache/spark/scheduler/TaskResultGetter.scala

```diff
@@ -51,14 +51,14 @@ private[spark] class TaskResultGetter(sparkEnv: SparkEnv, scheduler: TaskSchedul
         try {
           val (result, size) = serializer.get().deserialize[TaskResult[_]](serializedData) match {
             case directResult: DirectTaskResult[_] =>
-              if (!taskSetManager.canFetchMoreResults(serializedData.limit())) {
+              if (!taskSetManager.canFetchMoreResults(serializedData.remaining())) {
```

Comment:
This looks like it makes a unit test fail. I think you may have to check the size before the deserializer consumes the byte buffer?

This is overall looking good but we probably have to comb through these a little more to think through the implications.